### PR TITLE
Vishakh1/new db executor

### PR DESCRIPTION
**Pull request by vishakha-nervana**
New executor database class PR

### CODE REVIEW

File: ngraph_bridge/ngraph_executor_db.cc [hunk 1]
```diff
@@ -0,0 +1,125 @@
+
```
Comment (sindhu-nervana) [inline ngraph_bridge/ngraph_executor_db.cc]: Add copyright info for this file

File: ngraph_bridge/ngraph_executor_db.cc [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+  if (std::get<0>(io_tensors) < 0) {
+    return errors::Internal("No free tensor available");
+  }
+  return Status::OK();
+}
+
+void NGraphExecutorDB::RemoveItem(
+    std::string signature,
+    std::shared_ptr<ngraph::runtime::Executable>&
+        evicted_ng_exec)  // line no. 261, 262, 263
```
Comment (sindhu-nervana) [inline ngraph_bridge/ngraph_executor_db.cc]: are these numbers valid here? what do they mean?

File: test/test_executor_db.cpp [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+#include <atomic>
+#include <memory>
+#include "gtest/gtest.h"
+#include "tensorflow/core/common_runtime/optimization_registry.h"
+#include "tensorflow/core/graph/graph_constructor.h"
+#include "tensorflow/core/public/session.h"
+#include "test/test_utilities.h
```
Comment (sindhu-nervana) [inline test/test_executor_db.cpp]:
```suggestion
#include "test/test_utilities.h"
```

File: ngraph_bridge/ngraph_executor_db.cc (hunk 2 above)
Comment (vishakha-nervana) [reply ngraph_bridge/ngraph_executor_db.cc]: No they are not apparently my changes did not go through, will get rid of them

File: ngraph_bridge/CMakeLists.txt [hunk 4]
```diff
@@ -45,6 +45,7 @@ set(SRC
    ngraph_encapsulate_impl.cc
    ops/ngraph_ops.cc
    ngraph_encapsulate_op.cc
+   ngraph_executor_db.cc
```
Comment (sindhu-nervana) [inline ngraph_bridge/CMakeLists.txt]: Should add header and .cc file for bazel here - https://github.com/tensorflow/ngraph-bridge/blob/master/bazel/BUILD#L22

File: ngraph_bridge/ngraph_executor_db.cc [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+namespace ngraph_bridge {
+
+NGraphExecutorDB::~NGraphExecutorDB() {
+  m_executable_pipelined_tensors_map.clear();
+  m_ng_function_map.clear();
+  m_ng_exec_map.clear();
+}
+
+bool NGraphExecutorDB::MaybeGetNgExecutable(
+    std::string signature,
```
Comment (shresthamalik) [inline ngraph_bridge/ngraph_executor_db.cc]: Add const to the signature?

File: ngraph_bridge/ngraph_executor_db.cc [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    auto ng_exec_map_output_pair = m_ng_exec_map.emplace(signature, ng_exec);
+    if (ng_exec_map_output_pair.second) {
+      auto ng_func_map_output_pair =
+          m_ng_function_map.emplace(ng_exec, ng_function);
+      if (ng_func_map_output_pair.second) {
+        m_lru.push_front(signature);
+        auto it = m_executable_pipelined_tensors_map.find(
+            ng_exec);  // line no. 443-444
+
+        // IsPipelinedTensorsStoreAvailable
```
Comment (shresthamalik) [inline ngraph_bridge/ngraph_executor_db.cc]: We shouldnt be creating pipelined tensors here? They should be passed to this function like ng-function and exec

File: ngraph_bridge/ngraph_executor_db.cc [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+  UpdateLRU(signature);
+  return true;
+}
+
+void NGraphExecutorDB::AddItem(
+    std::string signature,
+    std::pair<std::shared_ptr<ngraph::runtime::Executable>,
+              std::shared_ptr<ngraph::Function>>
+        ng_exec_func,
+    std::shared_ptr<ngraph::runtime::Executable>& evicted_ng_exec, int depth) {
```
Comment (shresthamalik) [inline ngraph_bridge/ngraph_executor_db.cc]: Iterating @sayantan-nervana 's comment: move depth to the constructor

File: ngraph_bridge/ngraph_executor_db.cc [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+    lock_guard<mutex> lock(m_mutex);
+    const char* cache_depth_specified =
+        std::getenv("NGRAPH_TF_FUNCTION_CACHE_ITEM_DEPTH");
+    if (cache_depth_specified != nullptr) {
+      if (m_ng_exec_map.size() >= atoi(cache_depth_specified)) {
+        RemoveItem(m_lru.back(), evicted_ng_exec);
+      }
+    }
+    auto ng_exec_map_output_pair = m_ng_exec_map.emplace(signature, ng_exec);
+    if (ng_exec_map_output_pair.second) {
```
Comment (sindhu-nervana) [inline ngraph_bridge/ngraph_executor_db.cc]: Like we discussed, this is similar to `if(! MaybeGetNgExecutable)`. We could discuss more about this tomorrow

**Pull request closed without merging**